modules/javafx.graphics/src/main/java/javafx/scene/layout/VBox.java

```diff
@@ -457,77 +460,166 @@ private double[][] getAreaHeights(List<Node>managed, double width, boolean minim
+                } else {
+                    maxHeights[i] = -1;
+                }
             }
-        } else {
-            for (int i = 0, size = managed.size(); i < size; i++) {
-            final Node child = managed.get(i);
-            if (getVgrow(child) == priority) {
-                temp[i] = computeChildMaxAreaHeight(child, -1, getMargin(child), width);
-                adjustingNumber++;
-            } else {
-                temp[i] = -1;
+
+            // Adjust the children that are eligible in this phase and return early if the children
+            // fit the target height (so no second phase is required).
+            if (adjustHeightsWithinLimits(managed, currentHeights, maxHeights, targetHeight, adjustingNumber)) {
+                return;
             }
         }
-        }
+    }
 
-        double available = extraHeight; // will be negative in shrinking case
-        outer: while (Math.abs(available) > 1 && adjustingNumber > 0) {
-            final double portion = snapPortionY(available / adjustingNumber); // negative in shrinking case
-            for (int i = 0, size = managed.size(); i < size; i++) {
-                if (temp[i] == -1) {
+    /**
+     * Resizes the children heights to fit the target height, while taking into account the resize limits
+     * for each child (their minimum and maximum height). This method will be called once when shrinking,
+     * and may be called twice when growing.
+     *
+     * @param managed the managed children
+     * @param currentHeights the current children heights
+     * @param limitHeights the max or min heights for each child, depending on whether we are growing or shrinking;
+     *                     a value of -1 means the child cannot be resized
+     * @param targetHeight sum of children heights and spacing
+     * @param adjustingNumber a number that indicates how many children can be resized
+     * @return {@code true} if the children heights were successfully resized to fit the target height;
+     *         {@code false} otherwise
+     */
+    private boolean adjustHeightsWithinLimits(
+            List<Node> managed, double[] currentHeights, double[] limitHeights, double targetHeight, int adjustingNumber) {
+        double totalSpacing = (managed.size() - 1) * snapSpaceY(getSpacing());
+
+        // Current total height and current delta are two important numbers that we continuously
+        // update as this method converges towards a solution.
+        double currentTotalHeight = snapSpaceY(sum(currentHeights, managed.size())) + totalSpacing;
+        double currentDelta = targetHeight - currentTotalHeight;
+
+        // We repeatedly apply the following algorithm as long as we have space left to
+        // distribute (currentDelta), as well as children that are eligible to grow or
+        // shrink (adjustingNumber).
+        while ((currentDelta > Double.MIN_VALUE || currentDelta < -Double.MIN_VALUE) && adjustingNumber > 0) {
```

Comment:
same question: is it possible to enter an infinite loop?

I suspect the problem being solved here is the same as with the constrained column resize policy with non-integer scale JDK-8299753.  I was going to use some kind of simulated annealing algorithm there, perhaps it could be extracted into a generic utility and used in all three cases.